components/script/dom/document.rs

```diff
@@ -2399,10 +2401,98 @@ impl Document {
         }
     }
 
+    /// <https://drafts.csswg.org/cssom-view/#document-run-the-scroll-steps>
+    pub(crate) fn run_the_scroll_steps(&self, can_gc: CanGc) {
+        // Step 1.
+        // > Run the steps to dispatch pending scrollsnapchanging events for doc.
+        // TODO(#7673): Implement scroll snapping
+
+        // Step 2
+        // > For each item target in doc’s pending scroll event targets, in the order they
+        // > were added to the list, run these substeps:
+        for target in self.pending_scroll_event_targets.borrow().iter() {
+            // Step 2.1
+            // > If target is a Document, fire an event named scroll that bubbles at target.
+            if target.downcast::<Document>().is_some() {
+                target.fire_bubbling_event(Atom::from("scroll"), can_gc);
+            }
+
+            // Step 2.2
+            // > Otherwise, fire an event named scroll at target.
+            if target.downcast::<Document>().is_some() {
+                target.fire_bubbling_event(Atom::from("scroll"), can_gc);
+            }
+        }
+
+        // Step 3.
+        // > Empty doc’s pending scroll event targets.
+        self.pending_scroll_event_targets.borrow_mut().clear();
+
+        // Step 4.
+        // > Run the steps to dispatch pending scrollsnapchange events for doc.
+        // TODO(#7673): Implement scroll snapping
+    }
+
+    /// Step 2-4 of viewport scroll event.
```

Comment:
A better documentation would be to simply quote the spec: "Whenever a [viewport](https://drafts.csswg.org/css2/#viewport%E2%91%A0) gets scrolled (whether in response to user interaction or by an API), the user agent must run these steps"(keeping the link as well).

Reply:
Sounds more informative, thanks!